setup.py

```diff
@@ -6,21 +6,11 @@
 with open("openml/__version__.py") as fh:
     version = fh.readlines()[-1].split()[-1].strip("\"'")
 
-dependency_links = []
-
-try:
-    import numpy  # noqa: F401
-except ImportError:
-    print('numpy is required during installation')
-    sys.exit(1)
-
-try:
-    import scipy  # noqa: F401
-except ImportError:
-    print('scipy is required during installation')
+if len(sys.argv) > 1 and sys.argv[1] == 'install':
```

Comment:
Could you please briefly explain why this is necessary?

Reply:
It's not. It's just there for people who go by the `python setup.py install` routine, so they don't get the bad/failing install with no clear error message:

```
root@cb87886bef12:/openml-python# python3 setup.py install
/usr/lib/python3.5/distutils/dist.py:261: UserWarning: Unknown distribution option: 'project_urls'
  warnings.warn(msg)
running install
Checking .pth file support in /usr/local/lib/python3.5/dist-packages/
/usr/bin/python3 -E -c pass
TEST PASSED: /usr/local/lib/python3.5/dist-packages/ appears to support .pth files
running bdist_egg
 ....  skip 250 lines ...
copying openml.egg-info/top_level.txt -> build/bdist.linux-x86_64/egg/EGG-INFO
zip_safe flag not set; analyzing archive contents...
openml.__pycache__.testing.cpython-35: module references __file__
tests.test_extensions.__pycache__.test_functions.cpython-35: module MAY be using inspect.stack
tests.test_extensions.test_sklearn_extension.__pycache__.test_sklearn_extension.cpython-35: module references __file__
tests.test_examples.__pycache__.test_OpenMLDemo.cpython-35: module references __file__
creating dist
creating 'dist/openml-0.8.0-py3.5.egg' and adding 'build/bdist.linux-x86_64/egg' to it
removing 'build/bdist.linux-x86_64/egg' (and everything under it)
Processing openml-0.8.0-py3.5.egg
creating /usr/local/lib/python3.5/dist-packages/openml-0.8.0-py3.5.egg
Extracting openml-0.8.0-py3.5.egg to /usr/local/lib/python3.5/dist-packages
Adding openml 0.8.0 to easy-install.pth file

Installed /usr/local/lib/python3.5/dist-packages/openml-0.8.0-py3.5.egg
Processing dependencies for openml==0.8.0
Searching for numpy>=1.6.2
Reading https://pypi.python.org/simple/numpy/
Best match: numpy 1.16.2
Downloading https://files.pythonhosted.org/packages/cf/8d/6345b4f32b37945fedc1e027e83970005fc9c699068d2f566b82826515f2/numpy-1.16.2.zip#sha256=6c692e3879dde0b67a9dc78f9bfb6f61c666b4562fd8619632d7043fb5b691b0
Processing numpy-1.16.2.zip
Writing /tmp/easy_install-rvsl2scc/numpy-1.16.2/setup.cfg
Running numpy-1.16.2/setup.py -q bdist_egg --dist-dir /tmp/easy_install-rvsl2scc/numpy-1.16.2/egg-dist-tmp-f4wila54
Running from numpy source directory.
/tmp/easy_install-rvsl2scc/numpy-1.16.2/setup.py:390: UserWarning: Unrecognized setuptools command, proceeding with generating Cython sources and expanding templates
  run_build = parse_setuppy_commands()
/usr/lib/python3.5/distutils/dist.py:261: UserWarning: Unknown distribution option: 'python_requires'
  warnings.warn(msg)
/tmp/easy_install-rvsl2scc/numpy-1.16.2/numpy/distutils/system_info.py:638: UserWarning:
    Atlas (http://math-atlas.sourceforge.net/) libraries not found.
    Directories to search for the libraries can be specified in the
    numpy/distutils/site.cfg file (section [atlas]) or by setting
    the ATLAS environment variable.
  self.calc_info()
/tmp/easy_install-rvsl2scc/numpy-1.16.2/numpy/distutils/system_info.py:638: UserWarning:
    Blas (http://www.netlib.org/blas/) libraries not found.
    Directories to search for the libraries can be specified in the
    numpy/distutils/site.cfg file (section [blas]) or by setting
    the BLAS environment variable.
  self.calc_info()
/tmp/easy_install-rvsl2scc/numpy-1.16.2/numpy/distutils/system_info.py:638: UserWarning:
    Blas (http://www.netlib.org/blas/) sources not found.
    Directories to search for the sources can be specified in the
    numpy/distutils/site.cfg file (section [blas_src]) or by setting
    the BLAS_SRC environment variable.
  self.calc_info()
/bin/sh: 1: svnversion: not found
non-existing path in 'numpy/distutils': 'site.cfg'
/tmp/easy_install-rvsl2scc/numpy-1.16.2/numpy/distutils/system_info.py:638: UserWarning:
    Lapack (http://www.netlib.org/lapack/) libraries not found.
    Directories to search for the libraries can be specified in the
    numpy/distutils/site.cfg file (section [lapack]) or by setting
    the LAPACK environment variable.
  self.calc_info()
/tmp/easy_install-rvsl2scc/numpy-1.16.2/numpy/distutils/system_info.py:638: UserWarning:
    Lapack (http://www.netlib.org/lapack/) sources not found.
    Directories to search for the sources can be specified in the
    numpy/distutils/site.cfg file (section [lapack_src]) or by setting
    the LAPACK_SRC environment variable.
  self.calc_info()
/usr/lib/python3.5/distutils/dist.py:261: UserWarning: Unknown distribution option: 'define_macros'
  warnings.warn(msg)
/usr/lib/python3.5/distutils/dist.py:261: UserWarning: Unknown distribution option: 'python_requires'
  warnings.warn(msg)
Could not locate executable gfortran
Could not locate executable f95
Could not locate executable ifort
Could not locate executable ifc
Could not locate executable lf95
Could not locate executable pgfortran
Could not locate executable f90
Could not locate executable f77
Could not locate executable fort
Could not locate executable efort
Could not locate executable efc
Could not locate executable g77
Could not locate executable g95
Could not locate executable pathf95
Could not locate executable nagfor
don't know how to compile Fortran code on platform 'posix'
_configtest.c:1:5: warning: conflicting types for built-in function 'exp'
 int exp (void);
     ^
_configtest.o: In function `main':
/tmp/easy_install-rvsl2scc/numpy-1.16.2/_configtest.c:6: undefined reference to `exp'
collect2: error: ld returned 1 exit status
_configtest.c:1:5: warning: conflicting types for built-in function 'exp'
 int exp (void);
     ^
_configtest.c:1:24: fatal error: sys/endian.h: No such file or directory
compilation terminated.
_configtest.c: In function 'main':
_configtest.c:5:16: warning: variable 'test_array' set but not used [-Wunused-but-set-variable]
     static int test_array [1 - 2 * !(((long) (sizeof (npy_check_sizeof_type))) >= 0)];
                ^
_configtest.c: In function 'main':
_configtest.c:5:16: warning: variable 'test_array' set but not used [-Wunused-but-set-variable]
     static int test_array [1 - 2 * !(((long) (sizeof (npy_check_sizeof_type))) == 4)];
                ^
_configtest.c: In function 'main':
_configtest.c:5:16: warning: variable 'test_array' set but not used [-Wunused-but-set-variable]
     static int test_array [1 - 2 * !(((long) (sizeof (npy_check_sizeof_type))) >= 0)];
                ^
_configtest.c: In function 'main':
_configtest.c:5:16: warning: variable 'test_array' set but not used [-Wunused-but-set-variable]
     static int test_array [1 - 2 * !(((long) (sizeof (npy_check_sizeof_type))) == 8)];
                ^
_configtest.c: In function 'main':
_configtest.c:5:16: warning: variable 'test_array' set but not used [-Wunused-but-set-variable]
     static int test_array [1 - 2 * !(((long) (sizeof (npy_check_sizeof_type))) >= 0)];
                ^
_configtest.c: In function 'main':
_configtest.c:5:16: warning: variable 'test_array' set but not used [-Wunused-but-set-variable]
     static int test_array [1 - 2 * !(((long) (sizeof (npy_check_sizeof_type))) == 8)];
                ^
_configtest.c: In function 'main':
_configtest.c:5:16: warning: variable 'test_array' set but not used [-Wunused-but-set-variable]
     static int test_array [1 - 2 * !(((long) (sizeof (npy_check_sizeof_type))) >= 0)];
                ^
_configtest.c: In function 'main':
_configtest.c:5:16: warning: variable 'test_array' set but not used [-Wunused-but-set-variable]
     static int test_array [1 - 2 * !(((long) (sizeof (npy_check_sizeof_type))) == 16)];
                ^
_configtest.c: In function 'main':
_configtest.c:7:12: error: 'SIZEOF_LONGDOUBLE' undeclared (first use in this function)
     (void) SIZEOF_LONGDOUBLE;
            ^
_configtest.c:7:12: note: each undeclared identifier is reported only once for each function it appears in
_configtest.c: In function 'main':
_configtest.c:5:16: warning: variable 'test_array' set but not used [-Wunused-but-set-variable]
     static int test_array [1 - 2 * !(((long) (sizeof (npy_check_sizeof_type))) >= 0)];
                ^
_configtest.c: In function 'main':
_configtest.c:5:16: warning: variable 'test_array' set but not used [-Wunused-but-set-variable]
     static int test_array [1 - 2 * !(((long) (sizeof (npy_check_sizeof_type))) == 16)];
                ^
_configtest.c: In function 'main':
_configtest.c:5:16: warning: variable 'test_array' set but not used [-Wunused-but-set-variable]
     static int test_array [1 - 2 * !(((long) (sizeof (npy_check_sizeof_type))) >= 0)];
                ^
_configtest.c: In function 'main':
_configtest.c:5:16: warning: variable 'test_array' set but not used [-Wunused-but-set-variable]
     static int test_array [1 - 2 * !(((long) (sizeof (npy_check_sizeof_type))) == 32)];
                ^
_configtest.c: In function 'main':
_configtest.c:7:16: warning: variable 'test_array' set but not used [-Wunused-but-set-variable]
     static int test_array [1 - 2 * !(((long) (sizeof (npy_check_sizeof_type))) >= 0)];
                ^
_configtest.c: In function 'main':
_configtest.c:7:16: warning: variable 'test_array' set but not used [-Wunused-but-set-variable]
     static int test_array [1 - 2 * !(((long) (sizeof (npy_check_sizeof_type))) == 8)];
                ^
_configtest.c: In function 'main':
_configtest.c:7:16: warning: variable 'test_array' set but not used [-Wunused-but-set-variable]
     static int test_array [1 - 2 * !(((long) (sizeof (npy_check_sizeof_type))) >= 0)];
                ^
... many more minutes of error outputs ...
```

Reply:
Which reminds me I have to add/change the documentation regarding installing from the repo.

Reply:
Okay. I am very confused right now :D 
It seems the development branch installs just fine off `pip` (after removing the check for `scipy` and `numpy` from `setup.py` of course).
I would've sworn I had error messages before too, but maybe I got that mixed up with a `setup.py install`.
I went back and tried to install `scikit-learn` without first installing `numpy` and `scipy` first, and that also seems to work.
```
root@b40cfdd5527e:/# pip list
Package    Version
---------- -------
pip        19.0.3
setuptools 20.7.0
wheel      0.29.0
root@b40cfdd5527e:/# pip install scikit-learn
Collecting scikit-learn
  Downloading https://files.pythonhosted.org/packages/38/7f/7d21bd89a97603a14d3770ca912d4402c394a8e50642f9b8b5cc6867a61a/scikit_learn-0.20.3-cp35-cp35m-manylinux1_x86_64.whl (5.3MB)
    100% |################################| 5.3MB 1.1MB/s
Collecting scipy>=0.13.3 (from scikit-learn)
  Downloading https://files.pythonhosted.org/packages/f0/30/526bee2ce18c066f9ff13ba89603f6c2b96c9fd406b57a21a7ba14bf5679/scipy-1.2.1-cp35-cp35m-manylinux1_x86_64.whl (24.7MB)
    100% |################################| 24.7MB 838kB/s
Collecting numpy>=1.8.2 (from scikit-learn)
  Downloading https://files.pythonhosted.org/packages/e3/18/4f013c3c3051f4e0ffbaa4bf247050d6d5e527fe9cb1907f5975b172f23f/numpy-1.16.2-cp35-cp35m-manylinux1_x86_64.whl (17.2MB)
    100% |################################| 17.2MB 1.2MB/s
Installing collected packages: numpy, scipy, scikit-learn
Successfully installed numpy-1.16.2 scikit-learn-0.20.3 scipy-1.2.1
root@b40cfdd5527e:/# pip list
Package      Version
------------ -------
numpy        1.16.2
pip          19.0.3
scikit-learn 0.20.3
scipy        1.2.1
setuptools   20.7.0
wheel        0.29.0
root@b40cfdd5527e:/#
```
I don't really know what's going on here, but it seems my setup is at least not able to detect if `numpy` and `scipy` are actually required for `pip`.
I am currently testing from a docker container which is created from the ubuntu 16.04 image and python installed:
```
root@b40cfdd5527e:/# python3 -V
Python 3.5.2
root@b40cfdd5527e:/# cat /etc/lsb-release
DISTRIB_ID=Ubuntu
DISTRIB_RELEASE=16.04
DISTRIB_CODENAME=xenial
DISTRIB_DESCRIPTION="Ubuntu 16.04.5 LTS"
```

Reply:
About removing the install from the CI script. I agree, but I wasn't sure because it seems like a lot of dependencies are actually installed [here](https://github.com/openml/openml-python/blob/develop/ci_scripts/install.sh#L29). In particular I don't understand why `oslo.concurrency` and `scikit-learn` are installed at this step too, as they should be installed by `pip install -e '.[test]'`? And shouldn't the other dependencies also not just be specified under `extra_requires`?

Reply:
If possible they should all be listed as `install_requires` or `test_requires`.

Reply:
Resolving this as the way forward seems clear and it makes the rest of the PR easier to parse :)